struct_amd64.go

```diff
@@ -0,0 +1,172 @@
+// SPDX-License-Identifier: Apache-2.0
+// SPDX-FileCopyrightText: 2024 The Ebitengine Authors
+
+package purego
+
+import (
+	"math"
+	"reflect"
+)
+
+// https://gitlab.com/x86-psABIs/x86-64-ABI
+// Class determines where the 8 byte value goes.
+// Higher value classes win over lower value classes
+const (
+	NO_CLASS = 0b0000
+	SSE      = 0b0001
+	X87      = 0b0011 // long double not used in Go
+	INTEGER  = 0b0111
+	MEMORY   = 0b1111
+)
+
+func addStruct(v reflect.Value, numInts, numFloats, numStack *int, addInt, addFloat, addStack func(uintptr), keepAlive []interface{}) []interface{} {
+	if v.Type().Size() == 0 {
+		return keepAlive
+	}
+
+	var placeOnStack bool
+	var (
+		savedNumFloats = *numFloats
+		savedNumInts   = *numInts
+		savedNumStack  = *numStack
+	)
+	// if less than or equal to 64 bytes place in registers
+	if v.Type().Size() <= 8*8 {
+		placeOnStack = attemptPlaceRegisters(v, addFloat, addInt)
+	}
+	if placeOnStack {
+		// reset any values placed in registers
+		*numFloats = savedNumFloats
+		*numInts = savedNumInts
+		*numStack = savedNumStack
+		placeStack(v, addStack)
+	}
+	return keepAlive
+}
+
+func attemptPlaceRegisters(v reflect.Value, addFloat func(uintptr), addInt func(uintptr)) (placeOnStack bool) {
```

Comment:
* `tryPlaceRegisters` seems a better name.
* Returning `ok bool` seems more natural than `placeOnStack`. `ok` would be true when values are put on registers.